### PR TITLE
ReadAlongs Editor

### PR DESCRIPTION
**Pull request by roedoejet**
### PR Goal? 

The goal of this PR is many-fold:

- [x] Add @dhdaines' Editor code into the Studio-Web interface and angular-ify it
- [x] Allow the import of single-file HTML only
- [x] export editable.html in web bundle
- [x] click on word to sync wavesurfer
- [x] Add a demo of the ReadAlong and keep the labels and alignments in sync with the wavesurfer editor. This is so that images, translations, and slots can be edited as well
- [x] Preserve the alignment adjustment capability of the editor
- [x] Add a label-editing feature to adjust the labels
- [x] Allow the edited readalong to be downloaded into any other format

### New Issues
For another PR...

https://github.com/ReadAlongs/Studio-Web/issues/272

### Feedback sought? 

- Please spin up the editor `/editor` and try it out - try to break it
- Please also add translations for Spanish and French

### Priority? 

- Medium/High - this touches a ton of stuff, so we should work on getting this merged in before other work

### Tests added? 

- I have added spec tests for the new components/services, but don't have any e2e tests. I welcome suggestions of how to test this 

### How to test? 

Just spin up the server using the normal commands in the readme and go to localhost:4200/editor

### Confidence? 

medium/high (this is a large PR, and the way we handle assets internally is messy, so there are probably optimizations to be made here, but I can't seem to find any other bugs at the moment.)

#### Left TODOs

- [x] Needs translations for new UI components
- [x] BUG: editing alignments and then downloading doesn't always save the adjusted alignments
- [x] BUG: playing individual segments after editing alignments doesn't work 
- [x] BUG: audio not being added to web bundle

### Version change? 

Minor bump for WebComponent due to new methods. I've gone ahead and bumped us to 1.3.0 but before merging this into main we will need to publish on npm.

### CODE REVIEW

**Comment by dhdaines** [comment]
Wow!  This is cool ... though I'm not totally sure how it's supposed to work :)  I'll take a closer look later this weekend.

**Comment by roedoejet** [comment]
> Wow! This is cool ... though I'm not totally sure how it's supposed to work :) I'll take a closer look later this weekend.

Yea - I guess something similar to the tour in the main Studio should be added here, but that's a whole other thing :) maybe we can make an issue for that

**Comment by roedoejet** [comment]
> Wow! This is cool ... though I'm not totally sure how it's supposed to work :) I'll take a closer look later this weekend.

Eric and I had a chat and decided to _only_ accept offline HTML, because if we accept uploads of .readalong files, we run into a bunch of problems with images and slots. I'll bring it all up at the next meeting.

**Comment by joanise** [comment]
Do you know why adding `/editor` doesn't work on the PR preview link?

File: packages/studio-web/src/app/editor/editor.component.html [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      <div class="col-8">
+        <h2 i18n="Title for upload box" class="subtitle">
+          Select a single-file HTML ReadAlong file
+        </h2>
+        <label
+          i18n="Instructions for uploading editable RAS"
+          for="updateRAS"
+          class="form-label"
+          >Select a corresponding ReadAlong Studio temporary HTML file
+          (.html)</label
```
Comment (joanise) [inline packages/studio-web/src/app/editor/editor.component.html]: This message doesn't really make sense.
image

It's just supposed to expand on the "Select a single-file HTML ReadAlong file" title above. Maybe "Select a ReadAlong Studio .html file downloaded as \"Offline HTML\" from Studio-Web"? Well, I don't like this either, but something that expands on the title and guides the user to understand where/how they might have generated such a file before. Ideally, in a concise way, quite unlike what I'm doing here. 😄

**Comment by roedoejet** [comment]
> Do you know why adding `/editor` doesn't work on the PR preview link?

No, probably some interaction with github and angular routing

File: packages/studio-web/src/app/editor/segments/index.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
+import { Segment } from "./segment.js";
```
Comment (joanise) [inline packages/studio-web/src/app/editor/segments/index.js:1]: this `app/editor/segments/` directory needs a 3rd party Copyright notice, I'll add it in a PR on top of this PR.

**Review by joanise — COMMENTED** [review]
This looks good to me, in general. I tested it locally and it seems to work well. I've done the required French translations (though one will need to be redone for the string I suggest changing in English). I still need to do the Spanish translation. That'll come in a PR on top of this one, as well as a missing 3rd party Copyright notice.

File: packages/studio-web/src/app/editor/editor.component.html (hunk 1 above)
Comment (roedoejet) [reply packages/studio-web/src/app/editor/editor.component.html]: can we maybe just remove that line altogether?

Comment (joanise) [reply packages/studio-web/src/app/editor/editor.component.html]: Good point, it's redundant anyway.

Comment (joanise) [reply packages/studio-web/src/app/editor/editor.component.html]: If you do that, please take out its translation from the message file in each language.

Comment (roedoejet) [reply packages/studio-web/src/app/editor/editor.component.html]: yep, done.

**Comment by dhdaines** [comment]
> > Do you know why adding `/editor` doesn't work on the PR preview link?
> 
> No, probably some interaction with github and angular routing

Probably just that GitHub doesn't do Angular routing.  You can try the hack where you copy index.html to 404.html in deployment...

**Comment by roedoejet** [comment]
> > > Do you know why adding `/editor` doesn't work on the PR preview link?
> > 
> > 
> > No, probably some interaction with github and angular routing
> 
> Probably just that GitHub doesn't do Angular routing. You can try the hack where you copy index.html to 404.html in deployment...

trying now...

Update: nope, that didn't work. I think maybe GitHub looks for a top-level 404.html file, which won't help us here.

**Comment by roedoejet** [comment]
> > > > Do you know why adding `/editor` doesn't work on the PR preview link?
> > > 
> > > 
> > > No, probably some interaction with github and angular routing
> > 
> > 
> > Probably just that GitHub doesn't do Angular routing. You can try the hack where you copy index.html to 404.html in deployment...
> 
> trying now...
> 
> Update: nope, that didn't work. I think maybe GitHub looks for a top-level 404.html file, which won't help us here.

I suppose we could use # routing for the pr-previews? I might yagni this for now, or just make an issue to look into it...

**Comment by joanise** [comment]
> I suppose we could use # routing for the pr-previews? I might yagni this for now, or just make an issue to look into it...

I think it's reasonable to yagni this.

**Review by joanise — APPROVED** [review]
This is great, and I think ready to merge.
Tested locally, and it works well. Very nice to be able to fix both text and alignment.

**Review by dhdaines — APPROVED** [review]
Wow, yes, this is cool and works really well for me too.

**Comment by roedoejet** [comment]
Awesome! @joanise is going to update the CI workflows first, but we'll have to publish to npm before merging this in as it expects a 1.3.0 release of the web-component. Thanks @dhdaines for all the work on the original Editor!! I think maybe once we do some user testing we can archive that repo and add some documentation/links from the main page to the editor. I haven't done any testing to make sure that state is maintained when moving between the editor and the main studio for example though, so there's a bit more work to be done there.

**Comment by dhdaines** [comment]
I'll keep as a to-do to update the wavesurfer.js stuff for the newest version, possibly also to get it merged into wavesurfer.js - the main change is just allowing region boundaries to be snapped together and preventing regions from overlapping (something that ws.js suprisingly does not do, or at least didn't at the time...)

**Pull request merged**